src/Avalonia.Controls/DockPanel.cs

```diff
@@ -75,132 +92,186 @@ public bool LastChildFill
 
-            double parentWidth = 0;   // Our current required width due to children thus far.
-            double parentHeight = 0;   // Our current required height due to children thus far.
-            double accumulatedWidth = 0;   // Total width consumed by children.
-            double accumulatedHeight = 0;   // Total height consumed by children.
+            var leftSpacing = false;
+            var topSpacing = false;
+            var rightSpacing = false;
+            var bottomSpacing = false;
+            var childrenCount = LastChildFill ? Children.Count - 1 : Children.Count;
 
-            for (int i = 0, count = children.Count; i < count; ++i)
+            for (var index = 0; index < childrenCount; ++index)
             {
-                var child = children[i];
-                Size childConstraint;             // Contains the suggested input constraint for this child.
-                Size childDesiredSize;            // Contains the return size from child measure.
+                var child = Children[index];
+                var childConstraint = new Size(
+                    GetPositiveOrZero(availableSize.Width - accumulatedWidth),
+                    GetPositiveOrZero(availableSize.Height - accumulatedHeight));
 
-                // Child constraint is the remaining size; this is total size minus size consumed by previous children.
-                childConstraint = new Size(Math.Max(0.0, constraint.Width - accumulatedWidth),
-                                           Math.Max(0.0, constraint.Height - accumulatedHeight));
-
-                // Measure child.
                 child.Measure(childConstraint);
-                childDesiredSize = child.DesiredSize;
-
-                // Now, we adjust:
-                // 1. Size consumed by children (accumulatedSize).  This will be used when computing subsequent
-                //    children to determine how much space is remaining for them.
-                // 2. Parent size implied by this child (parentSize) when added to the current children (accumulatedSize).
-                //    This is different from the size above in one respect: A Dock.Left child implies a height, but does
-                //    not actually consume any height for subsequent children.
-                // If we accumulate size in a given dimension, the next child (or the end conditions after the child loop)
-                // will deal with computing our minimum size (parentSize) due to that accumulation.
-                // Therefore, we only need to compute our minimum size (parentSize) in dimensions that this child does
-                //   not accumulate: Width for Top/Bottom, Height for Left/Right.
-                switch (GetDock(child))
+                var childDesiredSize = child.DesiredSize;
+
+                switch (child.GetValue(DockProperty))
                 {
                     case Dock.Left:
+                        leftSpacing = true;
```

Comment:
You can merge the `Left/Right` and `Top/Bottom` cases (as it was before). There isn't really a need to have different flags since they've never checked individually. You can keep only two, one for `horizontal` and one for `vertical`.